src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -222,43 +228,135 @@ def _push_content_single_part(self) -> str:
+                    # We can write sample_tsv directly to file, but participant_tsv may contain duplicates, which
+                    # need to be removed prior to writing to file.
+                    if fname == 'participant.tsv':
+                        s = fobj.getvalue()
+                        # Remove duplicates from participant.tsv
+                        f.write('\r\n'.join(list(filter(None, list(set(s.split('\r\n')))))) + '\r\n')
+                    else:
+                        f.write(fobj.getvalue())
+            for tsv_file in data.keys():
+                copy(os.path.join(tsv_file_dir, tsv_file), data_path)
+            assert ['participant.tsv', 'sample.tsv'] == os.listdir(data_path)
+            bag = bdbag_api.make_bag(bag_path, update=True)  # update TSV checksums
+            assert bdbag_api.is_bag(bag_path)
+            bdbag_api.validate_bag(bag_path)
+            assert bdbag_api.check_payload_consistency(bag)
+
+            return bdbag_api.archive_bag(bag_path, 'zip')
+
+    def _iterate_hit_tsv(self, es_search_hit, writer):
         hit_dict = es_search_hit.to_dict()
         assert len(hit_dict['contents']['files']) == 1
         file = hit_dict['contents']['files'][0]
         file_fields = self._translate(file, 'contents.files')
+
         for bundle in hit_dict['bundles']:
             # FIXME: If a file is in multiple bundles, the manifest will list it twice. `hca dss download_manifest`
             # would download the file twice (https://github.com/DataBiosphere/azul/issues/423).
             writer.writerow(self._translate(bundle, 'bundles') + file_fields)
 
+    def _iterate_hit_bdbag(self, es_search_hit, sample_writer, participant_writer):
+        hit_dict = es_search_hit.to_dict()
+
+        assert len(hit_dict['contents']['files']) == 1
+        assert len(hit_dict['contents']['specimens']) == 1
+
+        specimen = one(hit_dict['contents']['specimens'])
+        cell_suspension = one(hit_dict['contents']['cell_suspensions'])
+        file = hit_dict['contents']['files'][0]
+
+        specimen_fields = self._translate(specimen, 'contents.specimens')
+        cell_suspension_fields = self._translate(cell_suspension, 'contents.cell_suspensions')
+        file_fields = self._translate(file, 'contents.files')
+
+        # Construct URL for file.
+        base_url = 'https://service.staging.explore.data.humancellatlas.org'
+        file_id = file['uuid']
+        version = file['version']
+        replica = 'gcp'
+        endpoint = f'repository/files/{file_id}?version={version}&replica={replica}'
+        fetch_url = [urljoin(base=base_url, url=endpoint)]
+
+        for bundle in hit_dict['bundles']:
+            sample_writer.writerow(chain(specimen_fields[0], specimen_fields[1], cell_suspension_fields[0],
+                                         self._translate(bundle, 'bundles'), file_fields + fetch_url))
+            participant_writer.writerow(specimen_fields[1])
```

Comment:
This implies that the participant TSV has as many rows as the sample TSV. That can't be right.

Reply:
Initially yes, but it's pruned for duplicates in [_create_zipped_bdbag](https://github.com/DataBiosphere/azul/blob/feature/mkrause/604-export-to-terra/src/azul/service/responseobjects/hca_response_v5.py#L297).